### PR TITLE
Better CI run

### PR DESCRIPTION
**Pull request by santagada**
I did a bunch of changes to make the continuous integration/development smoother

- use pytest-cov and pytest-flake8 (can run both locally easier)
- always run flake8 on test runs (it caches the results so only run for changed files)
- run xonsh in debug mode
- turn on branch coverage (slower but more complete)
- configure files to not do coverage
- ignore all errors present currently on the code base for flake8
- added a conftest.py to make python3 -m pytest and py.test to behave the same

I will send more pull requests or maybe update this one removing the flake8 errors one by one or all in a bunch.

I think it will make writting the contributing docs on #1430 much easier

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -4,8 +4,8 @@ python:
   - 3.5
   - "nightly"
 install:
-  - pip install -r requirements-tests.txt
+  - pip install --upgrade -r requirements-tests.txt
 script:
-  - coverage run --source xonsh -m py.test -q
+  - py.test --flake8 --cov=. --cov-report=term
```
Comment (laerus) [inline .travis.yml]: i think this should be `--cov=xonsh` then you can stop omitting tests, setup etc on .coveragerc

Comment (scopatz) [inline .travis.yml]: shouldn't we have equivalent changes in appveyor?

**Comment by scopatz** [comment]
Thanks a million for putting this in!

File: .travis.yml (hunk 1 above)
Comment (santagada) [inline .travis.yml]: I don't know if we shouldn't be harcoding exactly what we don't care about? maybe we want to have test for amalgamation that is outside, maybe tests for other examples and other things as well? if not I can totally change.

**Comment by santagada** [comment]
changed appveyor although I don't know much about it and I don't know where to look for results.

File: .travis.yml (hunk 1 above)
Comment (laerus) [inline .travis.yml]: or we should hard code what we really want to test, we can always use more `--cov` arguments if we test stuff outside the xonsh package dir at a later time, seems easier than adding lines in `ommit:` each time we add something new to the root xonsh dir. The decision is yours though and i'm happy with either way!

**Comment by scopatz** [comment]
I am happy with this @santagada. I leave it up to @laerus to merge though!

**Comment by laerus** [comment]
Hey @santagada, do we still need the empty `conftest.py` in the root xonsh dir?

**Comment by santagada** [comment]
@laerus yes unfortunately, so new developers can just run py.test and be happy. We could put a not on why it is there.

**Comment by santagada** [comment]
@laerus it would be good to merge soon as new code just broke flake8 and I had to allow even more errors on the run.

**Comment by laerus** [comment]
ty @santagada , merging this!

**Pull request merged**